src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -115,33 +115,119 @@ public static void Analyze(CSharpCompilation compilation, Symbol member, BoundNo
 
         protected override void Normalize(ref LocalState state)
         {
+            int oldNext = state.Capacity;
             state.EnsureCapacity(nextVariableSlot);
+            for (int slot = oldNext; slot < nextVariableSlot; slot++)
+            {
+                var value = !GetDefaultState(ref state, slot);
+                state[slot] = value;
+            }
+        }
+
+        private bool? GetDefaultState(ref LocalState state, int slot)
+        {
+            if (slot == 0)
+            {
+                return null;
+            }
+
+            var variable = variableBySlot[slot];
+            var symbol = variable.Symbol;
+
+            switch (symbol.Kind)
+            {
+                case SymbolKind.Local:
+                    return null;
+                case SymbolKind.Parameter:
+                    {
+                        var parameter = (ParameterSymbol)symbol;
+                        return (parameter.RefKind == RefKind.Out) ?
+                            null :
+                            parameter.Type.IsNullable;
+                    }
+                case SymbolKind.Field:
+                case SymbolKind.Property:
+                    return !IsContainerAssigned(ref state, variable.ContainingSlot) ?
+                        null :
+                        symbol.GetTypeOrReturnType().IsNullable;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(symbol.Kind);
+            }
+        }
+
+        private bool IsContainerAssigned(ref LocalState state, int containingSlot)
+        {
+            var variable = variableBySlot[containingSlot];
+            var containingSymbol = variable.Symbol;
+            var containingType = containingSymbol.GetTypeOrReturnType();
+            if (containingType.TypeKind == TypeKind.Class && (object)containingSymbol == MethodThisParameter)
+            {
+                return true;
+            }
+            var isNotNull = state[containingSlot];
+            return isNotNull != null;
```

Comment:
>return isNotNull != null; [](start = 12, length = 25)

This feels fragile because the state doesn't really track assigned state. Potentially we could be in a situation when we simply don't know if the container is null, or simply would like to suppress warnings for it.

Reply:
This does not differentiate between unassigned and null-oblivious. (When we need to support null-oblivious, we'll need to start tracking assigned state explicitly.) For now, `null` is used for unassigned which is necessary particularly for reference types since member slots are populated lazily. We could remove the `IsContainerAssigned` check for reference types (and remove `this.State[slot] = valueIsNotNull;` at line 473), but we'll no longer recognize unassigned reference type instances such as `c` in `UnassignedOutParameterClassField`:
```
class C
{
    object? F;
    static void G(out C c)
    {
        object o = c.F;
        c.F.ToString();
    }
}
```

Reply:
I propose we ignore whether values are definitely assigned and instead treat all values as having the declared nullability by default, and warn for null mismatches regardless of whether the values are assigned. The alternative is to merge `NullabilityWalker` and `DataFlowPass` (revert #22932) to track assignment and nullability in a single pass. The two passes seems independent and the reason to merge them would be solely to suppress null warnings from uninitialized variables.

Reply:
Removed. `null` is no longer treated as unassigned.